20220714-lookup-api.md

```diff
@@ -0,0 +1,462 @@
+
+It will be provided as two different API endpoints: one will provide a streaming endpoint and the other will implement the endpoint without streaming. These two different endpoints could be used interchangably, but clients can choose which one may be better for their integration environment.
+
+Consider a developer/client integrating with OpenFGA with the following authorization model:
+```
+type folder
+    relations
+        define viewer as self
+
+type document
+    relations
+        define viewer as self or editor or viewer from parent
+        define editor as self
+        define parent as self
+```
+and the following relationship tuples:
+
+| object         | relation | user           |
+|----------------|----------|----------------|
+| document:doc1  | viewer   | bob            |
+| document:doc2  | editor   | bob            |
+| document:doc3  | parent   | folder:folder1 |
+| folder:folder1 | viewer   | bob            |
+
+## StreamedLookup
+StreamedLookup will implement a grpc [server streaming RPC](https://grpc.io/docs/what-is-grpc/core-concepts/#server-streaming-rpc) endpoint.
+
+The endpoint will stream each object id the target user has the given relationship with as it is evaluated, and the server will continuing streaming results until either the lookup has been exhausted or the `streamed-lookup-deadline` has been reached (see [Configuration Changes](#configuration-changes)).
+
+Sample grpcurl command:
+```
+grpcurl -plaintext \
+  -d '{
+           "store_id":"<store>", 
+           "authorization_model_id":"<modelID>",
+           "object_type":"document", 
+           "relation":"viewer", 
+           "user":"bob"
+      }' \
+  localhost:8081 openfga.v1.OpenFGAService/StreamedLookup
+
+{
+    "object_id": "doc1",
+}
+{
+    "object_id": "doc2"
+}
+{
+    "object_id": "doc3"
+}
+```
+
+Sample curl command:
+```
+curl --request POST 'http://localhost:8080/stores/<storeID>/lookup:streamed' \
+--header 'Content-Type: application/json' \
+--data-raw '{
+    "object_type": "document",
+    "relation": "viewer",
+    "user": "bob",
+    "authorization_model_id": "<modelID>"
+}'
+
+{"result":{"object_id":"doc1"}}
+{"result":{"object_id":"doc2"}}
+{"result":{"object_id":"doc3"}}
+```
+
+## Lookup
+Lookup will implement a grpc [unary RPC](https://grpc.io/docs/what-is-grpc/core-concepts/#unary-rpc) endpoint.
+
+The endpoint will return a list of the object ids the target user has the given relationship with. The size of the list that is returned will be determined by the `lookup-max-results` config (see [Configuration Changes](#configuration-changes)). If no maximum is provided by that config, then Lookup will attempt to accumulate all of the object ids before returning them. 
+> ⚠️ without proper configuration of or use of the Lookup endpoint the server memory (and client memory) footprint could be profoundly impacted, because the list is accumulated in memory before it is returned.
```

Comment:
```suggestion
> ⚠️ Without proper configuration the Lookup endpoint, the server's memory footprint could be profoundly impacted, because the list is accumulated in memory before it is returned.
```

Reply:
I reworded this a bit. Let me know what you think with the updated one.